designs/sam-app-config.md

```diff
@@ -0,0 +1,201 @@
+SAM CLI App Level Config
+====================================
+
+
+What is the problem?
+--------------------
+
+Today users of SAM CLI need to invoke the CLI directly with all parameters supplied to its commands.
+
+for e.g: `sam build --use-container --debug`
+
+But often, during the lifecycle of building and deploying a serverless application. the same commands get run repeatedly to build, package and deploy, before solidifying into the final application. 
+
+These CLI commands are often long and have many changing parts.
+
+Have a look at the following series of workflows
+
+
+* `sam build --use-container --template ... --parameter-overrides=... --skip-pull-image --manifest ...`
+
+* `sam package --s3-bucket ... --template-file ... --output-template-file ... --s3-prefix ... --kms-key-id ...`
+
+* `sam deploy deploy --template-file ... --stack-name ... --capabilities ... --tags ... --parameter-overrides ... --kms-key-id ...`
+
+If this could be condensed into a series of workflows that look like
+
+* `sam build`
+* `sam package`
+* `sam deploy`
+
+That would be a huge user experience win.
```

Comment:
I'd add to the problem some more considerations:

* Dev/Beta/PreProd/Prod workflows may vary and need to be reflected somehow.
* There is a tie-in with `~/.aws/credentials` and `~/.aws/config` to respect, or at least to keep consistent in some way.
* A solution should allow for overriding of any defaults - explicit parameter choices should always be respected.

Reply:
Not necessarily. ~/.aws/credentials and ~/.aws/config are global files. Yes, explicit parameter will always be respected. I can add that under a section called tenets.

Reply:
Those are global config files, but do we want the same profile identifier to work for both? Why or why not? Let's make a conscious decision.

Reply:
We can always leave the door open for identifiers, but I'd prefer not to start with one. It may conflate expectations with aws profiles.

Reply:
+1 to start with simple, as long as we keep door open to add profiles in future if needed